### PR TITLE
User fonts, macOS font fallback and refactoring

### PR DESCRIPTION
**Pull request by algesten**
1. Config change so `alacritty.yml` have lists of fonts that works as fallbacks.
2. Moved some shared logic to `font/src/lib.rs` so that user fallbacks can work in a similar way regardless of platform.
3. Supports two different fallback strategies depending on platform:
  - macOS adds on system defined font fallbacks at the end of the user font lists.
  - freetype/fontconfig attempts to get a fontconfig fallback per rendered glyph when needed.
4. Tidy up in `darwin` to split out some core-graphics stuff.

---
This change is [Reviewable](https://reviewable.io/reviews/jwilm/alacritty/575)

### CODE REVIEW

**Comment by jwilm** [comment]
This is a pretty big PR and I'm on travel the rest of this week. Will for sure review this weekend if I don't get to it sooner.

Thanks again for working on this feature. As I mentioned before, this is one of the most critical outstanding features for Alacritty on macOS. 💯

**Comment by algesten** [comment]
@jwilm cool! yes, in the end I felt I wanted macOS and freetype/fontconfig follow similar code paths avoid code duplication – ended up being quite some refactoring.

**Comment by cpick** [comment]
I did a bit of Linux testing and noted my results here: https://github.com/jwilm/alacritty/issues/45#issuecomment-302243308

**Comment by jwilm** [comment]
Thanks @cpick :)

**Comment by algesten** [comment]
rebased off current master

**Comment by iology** [comment]
I tried viewing some CJK text with Vim, sometimes some characters are half rendered with builtin color scheme "shine" and my custom ones (but no problem with `colorscheme default`). Those same characters are always half rendered in the same session no matter where they are positioned. I can't find a solid method to reproduce the problem. Will report later.

Some pics:

Normal: normal Crippled: clipped

<details>
<summary>Click here to see more samples</summary>

<details><summary>Text</summary>
<pre>
公有领域(CC)

原书名:The Public Domain: Enclosing the Commons of the Mind
原作者:詹姆斯·博伊尔（James Boyle）
原语种:英语
原书版权:CC 协议

本书简介

某种程度上说，我们的音乐、文化、科学和经济的繁荣与否取决于受控制的创意和自由的创意之间的博弈，取决于知识产权和公共领域之间的精妙平衡。在《公有领域：划定知识的公地》一书中，作者詹姆斯·博伊尔向读者介绍了“公有领域”的概念，并描述了这个领域如今是如何受到版权、专利和商标法的威胁。通过一系列案例，作者解释了为什么一些基本的商业创意和乐谱，以及基因序列会被私有；为什么如果爵士乐在今日被发明的话将是违法的行为；为什么直到现在大多数19世纪的文化在法律意义上仍不能公有，以及为什么现今的法律政策可能会把互联网扼杀在摇篮里。

为了实践本书主题，作者通过知识共享协议（CC BY-NC-SA 3.0）发布该书，读者可在作者网站上下载到该书的英文版。

作者詹姆斯·博伊尔是美国杜克大学法学教授，“公共领域研究中心”的联合创始人，知识共享组织的创始人之一。本项目属于公益项目，译本将免费提供给读者。
</pre>
</details>

<details><summary>Pic. normal</summary>
normal
</details>
<details><summary>Pic. crippled</summary>
clipped
</details>

</details>

**Comment by iology** [comment]
Is there a way to auto detect the italic and bold styles of the normal fonts?

**Comment by algesten** [comment]
The font fallback is probably not to blame for the drawing glitch.

**Comment by algesten** [comment]
@iology regarding bold/italic good question!

I think the answer is that we can take a stab at it, but may not succeed. many fonts exist in various "boldness" and those names are not necessarily always the same. they can be things like: 

hairline, ultra thin, thin, regular, bold, black, fat 

i don't think these are normalized, but you may be able to take a stab at guessing the one level up. I.e. if your "normal" font is source code pro ultra thin a good guess for bold is "source code pro thin"

**Comment by iology** [comment]
Agreed. I generalize to a simple rule from what I see in the FontBook.app:

* Regular → Italic, Bold, Bold Italic
* {Style} → {Style} Italic (no {Style} Bold Italic)

Also it seems so often that some fonts don't even has Bold or Italic styles, so it is good if they can be rendered bold/italic with a general method, without switching to an unknown fallback font and breaking the style.

These two functionalites cannot be combined without bothering the users. So I would prefer the latter one, the dumb way, to force auto-rendering of the characters in bold and/or italic. Then users can just leave the lists for "italic" and "bold" empty.

**Comment by iology** [comment]
Oh, they can be combined. First try Regular→Italic/Bold/Bold Italic, if no then auto rendering. Don't know if the system APIs support it.

**Comment by iology** [comment]
I find an issue: how to get ***Bold+Italic***? e.g. in Markdown `***bold and italic***`

**Comment by jwilm** [comment]
Thanks for keeping this rebased. Starting to look through it now.

**Comment by algesten** [comment]
I guess it needs another rebase now, but maybe I wait until you looked through it.

**Comment by polyzen** [comment]
Perhaps once this is merged, it would be a good time for a [release](https://github.com/jwilm/alacritty/milestone/1). :)

**Comment by jwilm** [comment]
@polyzen The milestone you linked is actually from the January announcement ;). We'll have another release as some point. There are a few things that I would like to fix/improve prior to that, though. I'll try and get a roadmap posted at some point.

@algesten thanks for your patience WRT a review. This PR contains changes to config layout, macOS font fallback (the primary herein), architectural changes impacting the _font_ crate, and changes to FreeType/fontconfig loading as a result. After much consideration, I would like to narrow the scope of this PR at this time to simply the macOS fallback feature. Let me explain.

First, the architectural changes haven't gotten us much in terms of reusability between backends, and it may be limiting for some backends (eg fontconfig). The primary feature provided by the wrapper type is managing a list of fonts and running a fallback algorithm with it. That functionality could instead be implemented as a general utility class that backends can use as needed (macOS, in this case).

By taking this approach, each backend can implement its optimal fallback strategy. Fontconfig will be able to search a `FontSet` instead of rolling our own list-based fallback, and macOS can use the list-based approach.

Second, I think breaking out the config changes to a separate PR will allow this macOS fallback PR to get merged sooner. Properly making those config changes would mean the fontconfig fallback needs changes to manage a second FontSet and try searching that before the global fallback list.

In short, here's what I would like to see:
* Back out config format changes for a second PR
* Back out architectural changes
* This PR becomes simply macOS fallback using the system cascade lists.

What do you think? As always, I'm open to discussion.

**Comment by algesten** [comment]
@jwilm I agree. I do this tomorrow.

**Comment by algesten** [comment]
First PR is #593 which hopefully is non controversial.

**Comment by sodiumjoe** [comment]
should I open another issue for cascading fallback? I'm not really following all the details, but this branch works for me while current master (with #594 merged) only appears to work with some characters.

**Comment by algesten** [comment]
@sodiumjoe can you give me an example character that doesn't work in master?

**Comment by sodiumjoe** [comment]
```
❯
```

**Comment by sodiumjoe** [comment]
that's with

```yaml
font:
  # The normal (roman) font face to use.
  normal:
    family: "InconsolataGo Nerd Font"

  # The bold font face
  bold:
    family: "InconsolataGo Nerd Font"

  # The italic font face
  italic:
    family: "InconsolataGo Nerd Font"
```
on this branch it worked with
```yaml
font:  # The normal (roman) font face to use.
  normal:
    family: "InconsolataGo Nerd Font"
    - family: Menlo
    - family: Apple Symbols

  # The bold font face
  bold:
    - family: "InconsolataGo Nerd Font"
    - family: Apple Symbols

  # The italic font face
  italic:
    - family: "InconsolataGo Nerd Font"
    - family: Apple Symbols
```

**Comment by algesten** [comment]
Ok. The glyph for `❯` is actually missing in "InconsolataGo Nerd Font" and it is also missing in my favorite "Source Code Pro". However, it is present in "Menlo" :/

**Comment by sodiumjoe** [comment]
Yeah, I figured that's why the list fallback worked. Is this still going to happen?

**Comment by jwilm** [comment]
@sodiumjoe yep! We identified some issues with the original config provided fallbacks on freetype/fontconfig systems so backed it out before merging.

**Comment by sodiumjoe** [comment]
Ok, sweet, is there another ticket I can track?

**Comment by algesten** [comment]
I'm looking at it right now. There seems to be glyphs we're missing when we use the system call to configure fallbacks using the user provided font.

I'm toying with always basing the fallbacks on Menlo to ensure we get all glyphs. More thoughts on this to come.

**Comment by algesten** [comment]
Buy maybe we should track the problem in a new issue.

**Comment by sodiumjoe** [comment]
Made a ticket for this: #606
